src/service-override/theme.ts

```diff
@@ -33,7 +33,7 @@ type PartialIThemeExtensionPoint = Partial<IThemeExtensionPoint> & Pick<IThemeEx
 export default function getServiceOverride (): IEditorOverrideServices {
   return {
     ...getFileServiceOverride(),
-    [IThemeService.toString()]: new SyncDescriptor(StandaloneWorkbenchThemeService)
+    [IThemeService.toString()]: new SyncDescriptor(StandaloneWorkbenchThemeService, undefined, true)
```

Comment:
You said you want it to be eager but the last parameter seems to do the opposite: it indicate if the service supports delayed instantiation

Reply:
I, of course, meant the delayed instantiation. So we want this service to be instantiated not when it is resolved as a dependency of another service, but when it is actually used, this way we have a bit more time to write the settings file so theme id from settings would be picked up and all the logic loading the cached theme data would work.

It still feels sloppy this implementation and I'd prefer not counting on that we will most likely write settings file timely, but that's a start!

Reply:
Okay, I see the problem now, in vscode it indeed is eager and not delayed. Thing is, when I was testing cached values were not picked up and it was still flickering because `settings.json` was not yet written so changing it to delayed did help a ton and so far have been working fine. But as said, that's really a very reliable solution

Reply:
The main problem here is that we do not have a configuration service initialized with the data from provided `user:/settings.json` as our service init sequence does not wait for that to be done.

I think the best solution for that might be a way to provide all the configuration data before initializing services and teach our config service override to use that data instead of relying on read that from fileService.

Otherwise we'd need to implement custom services init sequence where we'd wait for certain services to async init first, which I presume is undesirable

Reply:
It seems the storage service needs to be initialized. to do so, you call use `registerServiceInitializeParticipant`

Then you can implement `doInitialize` and do whatever you need in there, right?

Reply:
> It seems the storage service needs to be initialized. to do so, you call use `registerServiceInitializeParticipant`
> 
> Then you can implement `doInitialize` and do whatever you need in there, right?

Let me expand on this:

In order to avoid theme flickering both these requirements must be met, ideally:
1. Storage data to be available **during** the construction of `WorkspaceThemeService`, in our case it's during `initialize()` call
2. `workbench.colorTheme` configuration value to also be available **during** the construction of `WorkspaceThemeService`

First requirement is gonna be covered once this PR is merged, but the second one is currently impossible to do because we can only write the configuration file content (i.e. making `workbench.colorTheme` value be readable from the configuration service) **after** services has been initialized. There's no way to make the content of `user:/settings.json` available to configuration service prior to initialization of `WorkspaceThemeService`.

And so right now the weak workaround is to make `WorkspaceThemeService` to support delayed instantiation so we have a chance to write settings file **after** initialization and **before** actual construction of `WorkspaceThemeService`.

Hope this explains, haha.

Right now im just passing my own configuration service as an override so I can guarantee availability of all needed user settings during the monaco services initialization, but this is far from ideal because, well, I literally just copied config service code straight from vscode to ensure that monaco can use that w/o any incompatibilities.

Reply:
Looking at web.main.ts, it seems that workspace/configuration and storage services are initialized first, then all other services. We probably need something similar here

Reply:
> Looking at web.main.ts, it seems that workspace/configuration and storage services are initialized first, then all other services. We probably need something similar here

Yes, that's exactly what I meant, that ideally we'd want to migrate to init strategy like that and away from how standalone services do it.

That can be done in future PRs and I don't think that should be a blocker for this. How do you think?

Reply:
If I understand correctly, this PR is a partial fix and you still need to hack to fulfill your needs?

Reply:
> If I understand correctly, this PR is a partial fix and you still need to hack to fulfill your needs?

It is, indeed, partial right now.

Okay, let me just remove this theme service delayed instantiation as it is just a bad solution anyway.

> A proper fix seems quite simple: we probably just need to have a `registerServiceInitializeParticipant` AND a `registerServicePreInitializeParticipant` or something similar, and use the second one for configuration and storage

Ideally, yes, configuration service being inited prior to everything else will fix the issue in it's fullness. With this PR you don't need to preinit storage tho as provider here provides data synchronously so you would wait for that data prior to services initialization anyway.